`libraft` and `pylibraft` API for CAGRA build and HNSW search

divyegala opened the pull request:
Closes #1772

lowener commented on python/pylibraft/pylibraft/neighbors/cagra_hnswlib.pyx:
```diff
@@ -0,0 +1,445 @@
+
+@auto_sync_handle
+@auto_convert_output
+def search(SearchParams search_params,
```
Add all these new functions to the pylibraft API documentation

lowener commented on cpp/include/raft/neighbors/detail/cagra_hnswlib.hpp:
```diff
@@ -0,0 +1,186 @@
+#include <future>
+#include <memory>
+#include <mutex>
+#include <omp.h>
```
Unused?

cjnolet commented on cpp/cmake/thirdparty/get_hnswlib.cmake:
```diff
@@ -27,11 +27,20 @@ function(find_and_configure_hnswlib)
                 WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/_deps/hnswlib-src
         )
     endif ()
+    # rapids_cpm_find(hnswlib ${PKG_VERSION}
```
Can we remove this until we have e solution?

cjnolet commented on cpp/cmake/thirdparty/get_hnswlib.cmake:
```diff
@@ -27,11 +27,20 @@ function(find_and_configure_hnswlib)
                 WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/_deps )
 
         message("SOURCE ${CMAKE_CURRENT_SOURCE_DIR}")
+        message("WORKING DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}")
```
Is this just for debugging?

cjnolet commented on cpp/include/raft/neighbors/cagra_hnswlib_types.hpp:
```diff
@@ -0,0 +1,70 @@
+#include <sys/types.h>
+#include <type_traits>
+
+namespace raft::neighbors::cagra_hnswlib {
```
Rather than a separate index, namespace, and file for cagra_hnsw, can't we just use the cagra index, provide a helper in raft::neighbors::hnsw to convert it to the HNSW index wrapper, and then provide a raft::neighbors::hnsw::search()? I understand using the cagra_HNSW algo name for the benchmarks but I think HNSW itself is more appropriate for the actual public api to do this.

cjnolet commented on cpp/include/raft/neighbors/detail/cagra/cagra_serialize.cuh:
```diff
@@ -103,6 +103,8 @@ void serialize_to_hnswlib(raft::resources const& res,
                           std::ostream& os,
                           const index<T, IdxT>& index_)
 {
+  // static_assert(std::is_same_v<IdxT, int> or std::is_same_v<IdxT, uint32_t>,
```
Can we move this to hnsw_serialize?

cjnolet commented on cpp/include/raft/neighbors/detail/cagra_hnswlib.hpp:
```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
```
I think raft/neighbors/detail/hnsw.hpp would be more appropriate. By this point we have removed cagra completely from the story.

cjnolet commented on cpp/include/raft/neighbors/hnswlib_types.hpp:
```diff
@@ -0,0 +1,101 @@
+#include <sys/types.h>
+#include <type_traits>
+
+namespace raft::neighbors::cagra_hnswlib {
```
Let's just use `raft::neighbors::hnsw` for the namespace here.

cjnolet commented on cpp/include/raft/neighbors/hnswlib_types.hpp:
```diff
@@ -0,0 +1,101 @@
+  auto get_index() const -> void const* override { return appr_alg_.get(); }
+
+ private:
+  std::unique_ptr<hnswlib::HierarchicalNSW<typename hnsw_dist_t<T>::type>> appr_alg_;
```
This is leaning hnswlib types through the public API. This should just be a pure virtual class and the impl should be in detail (and produced via a factory function).

divyegala replied:
The pure virtual class is elsewhere. This class should be in detail.

cjnolet replied:
No, the types in these files should not be leaking any other dependencies. This is where the virtual class (eg interface) belongs so that hnswlib is guaranteed never to leak outside of here.

divyegala replied:
Yes, you are absolutely right. I put this class in detail.

cjnolet commented on cpp/include/raft_runtime/neighbors/cagra.hpp:
```diff
@@ -27,48 +27,56 @@
+              raft::device_matrix_view<const T, int64_t, row_major> queries,           \
+              raft::device_matrix_view<IdxT, int64_t, row_major> neighbors,            \
+              raft::device_matrix_view<float, int64_t, row_major> distances);          \
+  void serialize_file(raft::resources const& handle,                                   \
```
I think these belong in dedicated hnsw API.

cjnolet commented on cpp/include/raft_runtime/neighbors/cagra_hnswlib.hpp:
```diff
@@ -0,0 +1,43 @@
+namespace raft::runtime::neighbors::cagra_hnswlib {
+
+#define RAFT_INST_CAGRA_HNSWLIB_FUNCS(T)                                      \
+  void search(raft::resources const& handle,                                  \
```
Let's just put this still in an hnsw.hpp file (and namespace)

cjnolet commented on cpp/src/raft_runtime/neighbors/cagra_hnswlib.cpp:
```diff
@@ -0,0 +1,49 @@
+
+#include <raft/neighbors/cagra_hnswlib.hpp>
+#include <raft_runtime/neighbors/cagra_hnswlib.hpp>
+
```
Same here- hnswlib.cpp

cjnolet commented on cpp/include/raft_runtime/neighbors/cagra_hnswlib.hpp:
```diff
@@ -0,0 +1,43 @@
+
+#include <raft/core/host_mdspan.hpp>
+#include <raft/core/resources.hpp>
+#include <raft/neighbors/cagra_hnswlib_types.hpp>
```
This is going to leak hnswlib types through the headers. Not good

divyegala replied:
That file does not include any hnswlib headers.

cjnolet reviewed (changes requested):
I think the general idea is getting there, but there's still a lot of api leakage and we have an opportunity here for a dedicated HNSW API (independent from hnswlib).

divyegala commented:
@cjnolet I addressed your suggestions and reworked the filenames/namespaces/structure to a pure `hnsw` namespace. There's one thing that I'd like your opinion on and something I would consider a nice-to-have. As you requested, I moved the serialization and de-serialization functions into `hnsw_serialize.cuh`, however, here's the problem:
1. `serialize` accepts a CAGRA index and needs to be CUDA compatible
2. `deserialize` returns an hnsw index and does not need to be CUDA compatible

As a result of putting these two functions in the same header, they need to be compiled in the same `.cu` translation unit because `hnswlib` is not true header-only: it does not have `inline` functions hence all headers that contains an hnswlib included header must be compiled in the same TU.

By splitting `deserialize` and `serialize` into two TUs, we will be able to:
1. Compile `serialize` in `hnswlib_serialize.cu`
2. Compile `deserialize` and `search` in `hnsw.cpp`

Another benefit of this splitting is that users will be able to build CPU-only RAFT and still be able to load hnsw indexes and search them. If this change is acceptable to you, then I will go ahead and make it.

lowener commented:
One last remark on my side: HNSW supports Filtering functions in `searchKnn`, we should also expose this.

divyegala commented:
@lowener I filed an [issue](https://github.com/rapidsai/raft/issues/2069) for filtering functions as a follow-on, this PR is pretty big already

divyegala commented on cpp/include/raft_runtime/neighbors/cagra.hpp:
```diff
@@ -27,48 +27,53 @@
+             raft::host_matrix_view<const T, int64_t, row_major> dataset)              \
+    ->raft::neighbors::cagra::index<T, IdxT>;                                          \
+                                                                                       \
+  void build_device(raft::resources const& handle,                                     \
```
Not sure, this change predates this PR

cjnolet commented:
/merge
